packages/labs/src/multi-select/multiSelect.tsx

```diff
@@ -0,0 +1,299 @@
+
+        return (
+            <Popover
+                autoFocus={false}
+                canEscapeKeyClose={true}
+                enforceFocus={false}
+                isOpen={this.state.isOpen}
+                position={Position.BOTTOM_LEFT}
+                {...popoverProps}
+                className={classNames(listProps.className, popoverProps.className)}
+                onInteraction={this.handlePopoverInteraction}
+                popoverClassName={classNames("pt-multi-select-popover", popoverProps.popoverClassName)}
+                popoverDidOpen={this.handlePopoverDidOpen}
+            >
+                <div
+                    onKeyDown={this.handleKeyDown}
+                    onKeyUp={this.state.isOpen && handleKeyUp}
+                >
+                    <TagInput
+                        inputProps={defaultInputProps}
+                        onRemove={this.handleTagRemove}
+                        {...tagInputProps}
+                        className={classNames("pt-multi-select", tagInputProps.className)}
+                        values={this.state.selectedItems.map(this.props.tagRenderer)}
+                    />
+                </div>
+                <div onKeyDown={handleKeyDown} onKeyUp={handleKeyUp}>
+                    <ul className={Classes.MENU} ref={listProps.itemsParentRef}>
+                        {this.renderItems(listProps)}
+                    </ul>
+                </div>
+            </Popover>
+        );
+    }
+
+    private renderItems({ activeItem, filteredItems, handleItemSelect }: IQueryListRendererProps<T>) {
+        const { itemRenderer, noResults } = this.props;
+
+        if (filteredItems.length === 0) {
+            return noResults;
+        }
+
+        return filteredItems.map((item, index) => itemRenderer({
+            index,
+            item,
+            handleClick: (e) => handleItemSelect(item, e),
+            isActive: item === activeItem,
+            isSelected: this.isItemSelected(item),
+        }));
+    }
+
+    private handleQueryChange = (event: React.FormEvent<HTMLInputElement>) => {
+        const query = event.currentTarget.value;
+        let nextState = { query };
+
+        if (query.length === 0 && this.props.openOnKeyDown) {
+            nextState = { ...nextState, isOpen: false };
+        }
+
+        this.setState(nextState);
+    }
+
+    private getSelectedItemIndex(item: T) {
+        return this.state.selectedItems.indexOf(item);
+    }
+
+    private isItemSelected(item: T) {
+        return this.getSelectedItemIndex(item) !== -1;
+    }
+
+    private selectItem(item: T) {
+        const { openOnKeyDown, resetOnSelect } = this.props;
+        let nextState = { selectedItems: [...this.state.selectedItems, item] };
+
+        if (openOnKeyDown) {
+            nextState = { ...nextState, isOpen: false };
+        }
+        if (resetOnSelect) {
+            nextState = { ...nextState, query: "" };
+        }
```

Comment:
this should be doable all at once, rather than creating up to three objects in one go. also you lose all type safety with your approach cuz it's just an `object`. 

something like:

```js
this.setState({
   selectedItems: ...
   isOpen: openOnKeyDown ? false : this.state.isOpen,
   query: resetOnSelect ? "" : this.state.query
});
```